### DISCUSSION TITLE
AD Update Parameter

### DISCUSSION THREAD

**Discussion by zichenwang01** [mitsuba-renderer/mitsuba3#816] [Q&A] [answered]
Hi,

I want to forward a single parameter to the rendered image and optimize the parameter. However, right now once I update the parameter, the gradients all become zero. Am I updating the parameter incorrectly?

    image = my_integrator.render(p)
    dr.set_grad(p, 1)
    dr.forward_to(p, image)
    print(dr.grad(image)) # works normally

    p += 0.001
    image = my_integrator.render(p)
    dr.set_grad(p, 1)
    dr.forward_to(p, image)
    print(dr.grad(image)) # grads all zero

**Comment by zichenwang01** [comment]
Btw, if I do not update the parameter with `p += 0.001`, then the second grad is correct

**Reply by njroussel** [reply]
Hi @zichenwang01 

Could you post a full reproducer? There are many possible causes to this issue. I'd say what most commonly happens here is that you're not explicitly calling `params.update()`.

**Reply by zichenwang01** [reply]
Indeed, I did not call `params.update()` because `p` in my setting is not inside `params` but instead a global variable. Also, I did not use an `optimizer` but instead wrote a simple gradient descent. 

        for epoch in range(100):
        total_loss = 0
        for i in range(1):
            img = sdf_integrator.render(scene=scene, params=params, sdf=sdf, 
                                    sensor=sensors[i], spp=16)
            
            loss = l2_loss(ref_imgs[i], img)
            print("loss: ", loss)
            
            dr.set_grad(loss, 1.0)
            # dr.backward(loss)
            dr.backward_to(p)
            total_loss += loss[0]
            
            p = p - 1e-2 * dr.grad(loss)
            params.update()
            print("p: ", p)

Here is a snippet of my custom integrator. No other places use `p`

        def sample(self, 
              scene:mi.Scene, sdf:SDF, sampler:mi.Sampler, 
              ray:mi.Ray3f, depth = 0, 
              active = True, **kwargs
        ):       
        # ray intersection with the scene 
        itx : SurfaceInteraction3f = scene.ray_intersect(ray)
        itx.p = dr.replace_grad(itx.p, Vector3f(0, p, 0)) # sphere move up
        # itx.p = dr.replace_grad(itx.p, Vector3f(p, 0, 0)) # sphere move right

The full reproducer contains a base integrator, a custom direct integrator, a scene XML, and an optimization file. I can provide these if needed!

**Reply by njroussel** [reply, marked as answer]
Hi @zichenwang01 

I think this just basic Python at play  After the `p = p - 1e-2 * dr.grad(loss)` the `p` variable in the body no longer matches the `p` in the integrator `sdf_integrator`. You're then trying to `backward_to(p)` which is the `p` defined in the loop body and not the integrator and is therefore completely detached from the loss computation (this will obviously give a 0 gradient) .
This is hard to write out clearly, I hope it was clear. Your issue should be solved by doing something like `sdf_integrator.p = p`. 

Without a full minimal reproducer I can't tell for sure.

**Reply by zichenwang01** [reply]
I see. Thanks!
